python/src/indexing.cpp

```diff
@@ -435,6 +463,8 @@ array mlx_get_item(const array& src, const nb::object& obj) {
     return mlx_get_item_int(src, nb::cast<nb::int_>(obj));
   } else if (nb::isinstance<nb::tuple>(obj)) {
     return mlx_get_item_nd(src, nb::cast<nb::tuple>(obj));
+  } else if (nb::isinstance<nb::list>(obj)) {
```

Comment:
That's great!

Reply:
So, you want me to change the code in a way that in `mlx_get_item`, we first convert the elements to be one of `ints, arrays, slices, ellipses and Nones` (and thus, lists are not an option from here. If we have a list, we covert it to `array`) and then, proceed with the logic provided in `mlx_get_item` as it is? How is it different from what we are currently doing? (Just checking if the object `is instance<nb::list>` and if so, we do the conversion (or normalisation) inside `mlx_get_item_list` function.
OR maybe I misunderstood your point

Reply:
I think you are referring to this pseudo-code:
```python
def __getitem__(self, indices):
    if isinstance(indices, tuple):
        indices = tuple(mx.array(x) if isinstance(x, list) else x for x in indices)
    return self.old_getitem(indices)
```

But I do not quite understand it. You mean, for list indexing, we put the lists inside a tuple and then do the indexing in such way? But why should we go in this way instead of full converting to an `mlx.array`?

Reply:
I think there may be a misunderstanding here. Multidimensional indexing is only performed using tuples.

When calling `x[a, b, c, d]` this really calls `x.__getitem__((a, b, c, d))` . Notice the double parentheses. The index passed to `__getitem__` will be a tuple. Passing a list does **not** perform multidimensional indexing. It indexes the 0-th dimension only!

So in the code-snippet above. `if not isinstance(indices, tuple)` then we only have one index for the 0-th dimension.

To give you an idea of the type of solution that we are looking for, the following implements the above python snippet and is a correct implementation of indexing using lists. Ideally we want this but without having to cast back and forth from C++ types to python types and without code duplication so that it covers all the setters as well.

```cpp
      .def("__getitem__", [](const array& src, nb::object indices) {
            if (nb::isinstance<nb::list>(indices)) {
              indices = nb::cast(array_from_list(indices, {}));
            } else if (nb::isinstance<nb::tuple>(indices)) {
              std::vector<nb::handle> new_indices;
              for (auto arg : nb::cast<nb::tuple>(indices)) {
                if (nb::isinstance<nb::list>(arg)) {
                  new_indices.push_back(nb::cast(array_from_list(arg, {})));
                } else {
                  new_indices.push_back(arg);
                }
              }
              indices = nb::tuple(nb::cast(new_indices));
            }
            return mlx_get_item(src, indices);
          }, nb::arg().none())

```